File: app/src/main/java/com/dmytry/microgeiger2/MicroGeiger2App.java
/*
    Copyright 2013 Dmytry Lavrov

    This file is part of MicroGeiger for Android.

    MicroGeiger is free software: you can redistribute it and/or modify it under the terms of the 
    GNU General Public License as published by the Free Software Foundation, either version 2 
    of the License, or (at your option) any later version.

    MicroGeiger is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along with MicroGeiger. 
    If not, see http://www.gnu.org/licenses/.
*/

package com.dmytry.microgeiger2;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioRecord;
import android.media.AudioTrack;
import android.media.MediaRecorder.AudioSource;
import android.media.MicrophoneInfo;
import android.os.SystemClock;
import android.preference.PreferenceManager;
import android.util.Log;

import java.io.IOException;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class MicroGeiger2App extends Application {
	private static final String TAG = "MicroGeiger";
	static final float running_avg_const=0.25f;

	public volatile int total_count=0;
	public final int min_clicks_in_queue=100;
	public final float min_smoothing_duration_sec=0.5f;

	public final int sample_rate=44100;
	public final int counters_update_rate=2;/// sample rate must be divisible by counters update rate

	public volatile long total_sample_count=0;
	
	public final int log_interval=6*sample_rate;/// logging interval in samples
	public java.util.Vector<Integer> counts_log=new java.util.Vector<Integer>();
	public int log_countdown=0, log_interval_click_count=0;
	
	public final int samples_per_update=sample_rate/counters_update_rate;

	//public volatile boolean changed=false;
	public volatile int change_count=0;

	boolean started=false;
	boolean connected=false;

	static public class IIRFilter{
		float running_avg=0;
		float get_value(float normalized_in){
			float v=normalized_in-running_avg;
			running_avg=running_avg*(1.0f-running_avg_const)+normalized_in*running_avg_const;
			return v;
		}
	}

	public class Counter{
		public int counts[];
		public int pos=0;		
		volatile public int count=0;
		public double scale=1.0;
		public String name;
		Counter(int n_counts, double scale_, String name_){
			counts=new int[n_counts];
			scale=scale_;
			name=name_;
		}
		public void push(int n){			
			pos=pos+1;
			if(pos>=counts.length)pos=0;
			int old_count=count;
			count-=counts[pos];
			counts[pos]=n;
			count+=n;
			if(count!=old_count)change_count++;
		}
		double getValue(){
			return scale*count;
		}
	}
	public volatile Counter counters[];

	static public class Click{
		public long time_in_samples=0;
	}

	public volatile Deque<Click> last_n_clicks=new LinkedList<Click>();

	void AppendClick(long time_in_samples){
		Click c=new Click();
		c.time_in_samples=time_in_samples;
		synchronized(last_n_clicks) {
			last_n_clicks.addLast(c);
		}
	}
	void TrimQueue(long time_in_samples) {
		long cutoff=time_in_samples-(long)(min_smoothing_duration_sec*sample_rate);
		synchronized(last_n_clicks){
			while(last_n_clicks.size() > min_clicks_in_queue && last_n_clicks.getFirst().time_in_samples<cutoff) {
				last_n_clicks.removeFirst();
			}
		}
	}

	float GetQueueCPM(){
		synchronized(last_n_clicks){
			if(last_n_clicks.size()==0)return 0;
			long sc=total_sample_count;
			long duration_in_samples=sc-last_n_clicks.getFirst().time_in_samples;
			if(duration_in_samples<=0)return 0;
			long count=last_n_clicks.size();
			return count*sample_rate*60.0f/duration_in_samples;
		}
	}

  
    public class Listener implements Runnable{
    	public volatile boolean do_stop=false;

		IIRFilter filter=new IIRFilter();

    	int current_offset=0;
		public short input_buffer[];
		short playback_buffer[];

		AudioRecord recorder;

		AudioTrack player;

		int dead_time=sample_rate/2000;
		double threshold=0.1;
		// High pass filter , peak is reached in 3 samples
		double rms_avg=0.0;
		double peak_meter=0.1;
		double peak_meter_decay=100.0/sample_rate;

		float click_volume=1.0f;
		int dead_countdown=0;
		int click_countdown=0;

		int click_duration=40;
		int click_beep_divisor=20;

		int sample_update_counter=0;
		int sample_count=0;

		final short getFromBufferAt(int i){
			i%=input_buffer.length;
			if(i<0)i+=input_buffer.length;
			return input_buffer[i];
		}

		// Circular buffer reading
		// Returns how many bytes were read
		int readIntoBuffer(AudioRecord recorder, int wanted_read, boolean blocking) { // int read_size = recorder.read(input_buffer,0,data_size, AudioRecord.READ_NON_BLOCKING);
			// sanitize
			if(wanted_read>input_buffer.length)wanted_read=input_buffer.length;

			int max_read_size=wanted_read;
			// Are we wrapping around the buffer?
			Boolean partial=max_read_size > input_buffer.length-current_offset;
			if(partial) {
				max_read_size=input_buffer.length-current_offset;
			}
			int bytes_read=recorder.read(input_buffer, current_offset, max_read_size,  blocking ? AudioRecord.READ_BLOCKING : AudioRecord.READ_NON_BLOCKING);

			int new_offset=current_offset+bytes_read;
			// Wraparound
			if(new_offset>=input_buffer.length){
				new_offset=0;
				if(partial){// Wraparound and read was partial, need another read
					bytes_read+=recorder.read(input_buffer, new_offset, wanted_read-bytes_read, blocking ? AudioRecord.READ_BLOCKING : AudioRecord.READ_NON_BLOCKING);
				}
			}
			return bytes_read;
		}

		// Blocking read for min_read plus queue emptying up to max_read
		int readAtLeast(AudioRecord recorder, int min_read, int max_read){
			int result=readIntoBuffer(recorder, min_read, true);
			if(result<max_read) {
				result+=readIntoBuffer(recorder, max_read-result, false);
			}
			return result;
		}


		@Override
		public void run() {	

			int record_min_buffer_size=AudioRecord.getMinBufferSize(sample_rate, AudioFormat.CHANNEL_OUT_FRONT_LEFT | AudioFormat.CHANNEL_OUT_FRONT_RIGHT, AudioFormat.ENCODING_PCM_16BIT);
			int play_min_buffer_size=AudioTrack.getMinBufferSize(sample_rate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);
			int min_buffer_size=Math.max(record_min_buffer_size, play_min_buffer_size);
			
			int data_size=Math.max(sample_rate/4, min_buffer_size);
			// Store about 160 seconds in circular buffer for the waveform viewer, power of 2 so that wraparound of sample numbers isn't a problem
			input_buffer=new short[Math.max(data_size,1024*1024)];
			current_offset=0;
			playback_buffer =new short[data_size*2];

			int recorder_buffer_size_bytes=4*Math.max(sample_rate/10, min_buffer_size);
			try {
				recorder = new AudioRecord(AudioSource.DEFAULT, sample_rate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, recorder_buffer_size_bytes);
			}catch(SecurityException ex) {
				Log.d(TAG, "No audio permission");
				return;
			}

			player = new AudioTrack(AudioManager.STREAM_RING,
					sample_rate, /* AudioFormat.CHANNEL_OUT_MONO */ AudioFormat.CHANNEL_OUT_FRONT_LEFT | AudioFormat.CHANNEL_OUT_FRONT_RIGHT,
	                AudioFormat.ENCODING_PCM_16BIT, 4*data_size,
	                AudioTrack.MODE_STREAM);

			Log.d(TAG, "Output channels: "+player.getChannelCount());
	        player.play();

			try{
			    while(!do_stop){
					getParametersFromConfig();
					if (recorder.getState()== AudioRecord.STATE_INITIALIZED){ // check to see if the recorder has initialized yet.
			            if (recorder.getRecordingState()== AudioRecord.RECORDSTATE_STOPPED){
			                 recorder.startRecording();
			            }else{
			            	// This is slow for some reason, todo: use events instead
							boolean is_peripheral = checkIfPeripheralIsConnected();

							if( is_peripheral ){
			            		if(!connected)change_count++;
			            		connected=true;
								long start_t_ = SystemClock.elapsedRealtime();
				            	//int read_size = recorder.read(input_buffer,0,data_size, AudioRecord.READ_NON_BLOCKING);

								// read at least 0.1 seconds of audio
								int read_size = readAtLeast(recorder, 4410, input_buffer.length);//readIntoBuffer(recorder, input_buffer.length);

								long end_t_ = SystemClock.elapsedRealtime();
								Log.d(TAG, "Read: "+read_size+" duration="+(end_t_-start_t_)+" t="+end_t_);
				            	int old_total_count=total_count;
				            	processInputDataAndGenerateClicks(read_size, playback_buffer);
								TrimQueue(total_sample_count);
				            	if(old_total_count!=total_count){
									change_count++;
				            	}
								playOutputSounds(read_size);
							}else{/// wired headset is not on
			            		if(connected)change_count++;
			            		connected=false;
			            		if(recorder!=null) {
									recorder.stop();
									recorder.release();
								}
								Thread.sleep(500);
								try {
									recorder = new AudioRecord(AudioSource.DEFAULT, sample_rate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, recorder_buffer_size_bytes);
								}catch(SecurityException ex) {
									Log.d(TAG, "No audio permission");
									return;
								}
			            		Thread.sleep(500);
			            	}
				    	}
			    	}else{
				    	Log.d(TAG, "failed to initialize audio");
				    	Thread.sleep(5000);
				    }
			    }
			} catch (InterruptedException e){
			}
		    finally{
		    	recorder.stop();
		    	recorder.release();
		    	player.stop();
		    	player.release();
		    }
		}

		private void playOutputSounds(int read_size) {
			if(click_volume>0.001) {
				long start_t= SystemClock.elapsedRealtime();
				int how_much_to_write= read_size *2;
				// hack to reduce amount of buffering
				// read was more than 1/10th of a second, skip some writing
				if(read_size >4410) {
					how_much_to_write-=16;
				}
				how_much_to_write-=2;
				int wrote_size=0;
				if(how_much_to_write>0) {
					wrote_size = player.write(playback_buffer, 0, how_much_to_write);// , AudioTrack.WRITE_NON_BLOCKING
				}
				long end_t= SystemClock.elapsedRealtime();
				Log.d(TAG, "Time to write: "+(end_t-start_t));
			}
		}

		private void processInputDataAndGenerateClicks(int read_size, short[] playback_data) {
			for(int i = 0; i < read_size; ++i, ++current_offset){
				if(current_offset>=input_buffer.length)current_offset=0;
				total_sample_count++;
				if(dead_countdown>0){
					dead_countdown--;
				}
				if(i*2+1<playback_data.length) {
					if (click_countdown > 0) {
						click_countdown--;
						playback_data[i * 2] = (short) (Math.exp((click_volume - 1.0) * Math.log(10000)) * ((click_countdown / click_beep_divisor) % 2 == 1 ? 32767 : -32767));
						playback_data[i * 2 + 1] = playback_data[i * 2];
					} else {
						//playback_data[i]=0;
						// test beep
						//short beep=(short)((total_sample_count/40)%2 == 1 ? 1000:-1000);
						short beep = 0;
						playback_data[i * 2] = beep;
						playback_data[i * 2 + 1] = beep;
					}
				}
				sample_update_counter++;
				if(sample_update_counter>=samples_per_update){
					for(int j=0;j<counters.length;++j){
						counters[j].push(sample_count);
					}
					sample_update_counter=0;
					sample_count=0;
					//Log.d(TAG, "got a sample");
				}
				float raw_v=input_buffer[current_offset]*(1.0f/32768.0f);
				float v=filter.get_value(raw_v);

				if(/* v>threshold || */ v<-threshold){
					if(dead_countdown<=0){
						total_count++;
						sample_count++;
						log_interval_click_count++;
						dead_countdown=dead_time;
						if(click_countdown<=0)click_countdown=click_duration;
						AppendClick(total_sample_count);
						TrimQueue(total_sample_count);
					}
				}
				if(log_countdown<=0){
					log_countdown=log_interval;
					counts_log.add(log_interval_click_count);
					log_interval_click_count=0;
				}
				log_countdown--;
			}
		}

		private boolean checkIfPeripheralIsConnected() {
			boolean is_peripheral=false;
			boolean checked_peripheral=false;
			try{
				if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
					List<MicrophoneInfo> microphones=recorder.getActiveMicrophones();
					for(MicrophoneInfo m: microphones) {
						int l=m.getLocation();
						if(l==MicrophoneInfo.LOCATION_UNKNOWN || l==MicrophoneInfo.LOCATION_PERIPHERAL) {
							is_peripheral=true;
						}
					}
					checked_peripheral=true;
				}
			}catch(IOException e) {
				Log.d(TAG, "Failed to query connected microphones");
			}
			// if we were unable to determine if microphone is peripheral, fallback to wired headset check
			if(!checked_peripheral) {
				is_peripheral=((AudioManager) getApplicationContext().getSystemService(Context.AUDIO_SERVICE)).isWiredHeadsetOn();
			}
			return is_peripheral;
		}

		private void getParametersFromConfig() {
			SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
			try{
				threshold=Double.parseDouble(prefs.getString("threshold", ""));
			}catch(NumberFormatException e){
			}
			try{
				dead_time=(int)(0.001*sample_rate*Double.parseDouble(prefs.getString("dead_time", "")));
			}catch(NumberFormatException e){
			}
			try{
				click_volume=Float.parseFloat(prefs.getString("click_volume", "1.0"));
			}catch(NumberFormatException e){
			}
		}
	}
    
    Listener listener;
    Thread listener_thread;
    
    void init_counters(){
    	counters=new Counter[4];
    	counters[0]=new Counter(counters_update_rate*5, 12.0, " CPM in last 5 sec");
    	counters[1]=new Counter(counters_update_rate*30, 2.0, " CPM in last 30 sec");
    	counters[2]=new Counter(counters_update_rate*120, 0.5, " CPM in last 2 min");
    	counters[3]=new Counter(counters_update_rate*600, 0.1, " CPM in last 10 min");
    }
    
    void start(){
    	if(!started){	    	
	    	init_counters();
	    	if(listener_thread==null){
		        listener=new Listener();
		        listener_thread=new Thread(listener);
		        listener_thread.start();
	        }
	    	started=true;
    	}
    }
    void reset(){
    	total_count=0;
    	init_counters();
		change_count++;
    }
    void stop(){
    	if(listener!=null){
    		listener.do_stop=true;
    		listener_thread.interrupt();
    		try {
				listener_thread.join();
			} catch (InterruptedException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
    		listener_thread=null;
    	}  
    	started=false;
    }
    @Override
	public void onCreate() {    	
        super.onCreate();   
        start();
    }
    @Override
    public void onTerminate() {    	
        super.onTerminate();
        stop();
    }
}

File: app/src/main/java/com/dmytry/microgeiger2/MicroGeiger2App.kt
/*
    Copyright 2013 Dmytry Lavrov

    This file is part of MicroGeiger for Android.

    MicroGeiger is free software: you can redistribute it and/or modify it under the terms of the 
    GNU General Public License as published by the Free Software Foundation, either version 2 
    of the License, or (at your option) any later version.

    MicroGeiger is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along with MicroGeiger. 
    If not, see http://www.gnu.org/licenses/.
*/
package com.dmytry.microgeiger2

import android.app.Application
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.media.*
import android.media.MediaRecorder.AudioSource
import android.os.Build
import android.os.SystemClock
import android.preference.PreferenceManager
import android.util.Log
import java.util.*
import kotlin.math.max

class MicroGeiger2App : Application() {
    @JvmField
    @Volatile
    var totalCount = 0
    val minClicksInQueue = 100

    // Minimum queue length in seconds
    val minSmoothingDurationSec = 0.5f
    val sampleRate = 44100

    @Volatile
    var totalSampleCount: Long = 0
    val logInterval = 6 * sampleRate /// logging interval in samples
    var countsLog = Vector<Int>()
    var logCountdown = 0
    var logIntervalClickCount = 0

    @JvmField
    @Volatile
    var changeCount = 0
    var started = false

    @JvmField
    var connected = false

    class IIRFilter {
        var runningAvg = 0f
        fun getValue(normalized_in: Float): Float {
            val v = normalized_in - runningAvg
            runningAvg = runningAvg * (1.0f - running_avg_const) + normalized_in * running_avg_const
            return v
        }
    }

    class Click {
        @JvmField
        var timeInSamples: Long = 0
    }

    @JvmField
    @Volatile
    var lastNClicks: Deque<Click> = LinkedList()
    fun appendClick(time_in_samples: Long) {
        val c = Click()
        c.timeInSamples = time_in_samples
        synchronized(lastNClicks) { lastNClicks.addLast(c) }
    }

    fun trimQueue(time_in_samples: Long) {
        val cutoff = time_in_samples - (minSmoothingDurationSec * sampleRate).toLong()
        synchronized(lastNClicks) {
            while (lastNClicks.size > minClicksInQueue && lastNClicks.first.timeInSamples < cutoff) {
                lastNClicks.removeFirst()
            }
        }
    }

    fun getQueueCPM(): Float {
        synchronized(lastNClicks) {
            if (lastNClicks.size == 0) return 0.0f
            val sc = totalSampleCount
            val duration_in_samples = sc - lastNClicks.first.timeInSamples
            if (duration_in_samples < 0) {
                lastNClicks.clear()
                return 0.0f
            }
            if (duration_in_samples == 0L) return 0.0f
            val count = lastNClicks.size.toLong()
            return count * sampleRate * 60.0f / duration_in_samples
        }
    }

    inner class Listener : Runnable, AudioRouting.OnRoutingChangedListener,
        OnSharedPreferenceChangeListener {
        @Volatile
        var doStop = false
        var filter = IIRFilter()
        var currentOffset = 0
        var inputBuffer = ShortArray(1024 * 1024)
        var playbackBuffer = ShortArray(8000)
        var recorder: AudioRecord? = null
        var player: AudioTrack? = null

        @JvmField
        @Volatile
        var deadTime = sampleRate / 2000

        @JvmField
        @Volatile
        var threshold = 0.1

        // High pass filter , peak is reached in 3 samples
        @Volatile
        var clickVolume = 1.0f
        var deadCountdown = 0
        var clickCountdown = 0
        var clickDuration = 40
        var clickBeepDivisor = 20
        var sampleCount = 0

        @Volatile
        var isPeripheral = false
        fun getFromBufferAt(in_i: Int): Short {
            var i = in_i
            i %= inputBuffer.size
            if (i < 0) i += inputBuffer.size
            return inputBuffer[i]
        }

        // Circular buffer reading
        // Returns how many bytes were read
        private fun readIntoBuffer(
            recorder: AudioRecord,
            wanted_read: Int,
            blocking: Boolean
        ): Int { // int read_size = recorder.read(input_buffer,0,data_size, AudioRecord.READ_NON_BLOCKING);
            // sanitize
            var wanted_read = wanted_read
            if (wanted_read > inputBuffer.size) wanted_read = inputBuffer.size
            var max_read_size = wanted_read
            // Are we wrapping around the buffer?
            val partial = max_read_size > inputBuffer.size - currentOffset
            if (partial) {
                max_read_size = inputBuffer.size - currentOffset
            }
            var bytes_read = recorder.read(
                inputBuffer,
                currentOffset,
                max_read_size,
                if (blocking) AudioRecord.READ_BLOCKING else AudioRecord.READ_NON_BLOCKING
            )
            var new_offset = currentOffset + bytes_read
            // Wraparound
            if (new_offset >= inputBuffer.size) {
                new_offset = 0
                if (partial) { // Wraparound and read was partial, need another read
                    bytes_read += recorder.read(
                        inputBuffer,
                        new_offset,
                        wanted_read - bytes_read,
                        if (blocking) AudioRecord.READ_BLOCKING else AudioRecord.READ_NON_BLOCKING
                    )
                }
            }
            return bytes_read
        }

        // Blocking read for min_read plus queue emptying up to max_read
        private fun readAtLeast(recorder: AudioRecord, min_read: Int, max_read: Int): Int {
            var result = readIntoBuffer(recorder, min_read, true)
            if (result < max_read) {
                result += readIntoBuffer(recorder, max_read - result, false)
            }
            return result
        }

        override fun run() {
            Log.d(TAG, "Run")
            val sharedPref = PreferenceManager.getDefaultSharedPreferences(applicationContext)
            sharedPref.registerOnSharedPreferenceChangeListener(this)
            parametersFromConfig()
            val record_min_buffer_size = AudioRecord.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_OUT_FRONT_LEFT or AudioFormat.CHANNEL_OUT_FRONT_RIGHT,
                AudioFormat.ENCODING_PCM_16BIT
            )
            val play_min_buffer_size = AudioTrack.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT
            )
            val min_buffer_size = Math.max(record_min_buffer_size, play_min_buffer_size)

            val data_size = max(sampleRate / 4, min_buffer_size)
            // Store about 160 seconds in circular buffer for the waveform viewer, power of 2 so that wraparound of sample numbers isn't a problem
            inputBuffer = ShortArray(Math.max(data_size, 1024 * 1024))
            currentOffset = 0
            totalSampleCount = 0
            lastNClicks.clear()
            playbackBuffer = ShortArray(data_size * 2)
            val recorder_buffer_size_bytes = 4 * max(sampleRate / 10, min_buffer_size)
            try {
                recorder = AudioRecord(
                    AudioSource.DEFAULT,
                    sampleRate,
                    AudioFormat.CHANNEL_IN_MONO,
                    AudioFormat.ENCODING_PCM_16BIT,
                    recorder_buffer_size_bytes
                )
                recorder?.addOnRoutingChangedListener(this, null)
            } catch (ex: SecurityException) {
                Log.d(TAG, "No audio permission")
                return
            }
            if (recorder == null) return
            player = AudioTrack(
                AudioManager.STREAM_RING,
                sampleRate,  /* AudioFormat.CHANNEL_OUT_MONO */
                AudioFormat.CHANNEL_OUT_FRONT_LEFT or AudioFormat.CHANNEL_OUT_FRONT_RIGHT,
                AudioFormat.ENCODING_PCM_16BIT,
                4 * data_size,
                AudioTrack.MODE_STREAM
            )
            if (player == null){
                recorder?.stop()
                recorder?.release()
                return
            }
            Log.d(TAG, "Output channels: " + player?.channelCount)
            player?.play()
            try {
                isPeripheral = checkIfPeripheralIsConnected()
                while (!doStop) {
                    //getParametersFromConfig();
                    if (recorder?.state == AudioRecord.STATE_INITIALIZED) { // check to see if the recorder has initialized yet.
                        if (recorder?.recordingState == AudioRecord.RECORDSTATE_STOPPED) {
                            recorder?.startRecording()
                        } else {
                            // This is slow for some reason, todo: use events instead
                            if (isPeripheral) {
                                if (!connected) changeCount++
                                connected = true
                                val start_t_ = SystemClock.elapsedRealtime()
                                //int read_size = recorder.read(input_buffer,0,data_size, AudioRecord.READ_NON_BLOCKING);

                                // read at least 0.1 seconds of audio
                                val read_size = readAtLeast(
                                    recorder!!,
                                    4410,
                                    inputBuffer.size
                                ) //readIntoBuffer(recorder, input_buffer.length);
                                val end_t_ = SystemClock.elapsedRealtime()
                                //Log.d(TAG, "Read: "+read_size+" duration="+(end_t_-start_t_)+" t="+end_t_);
                                val old_total_count = totalCount
                                processInputDataAndGenerateClicks(read_size, playbackBuffer)
                                trimQueue(totalSampleCount)
                                if (old_total_count != totalCount) {
                                    changeCount++
                                }
                                playOutputSounds(read_size)
                            } else { /// wired headset is not on
                                if (connected) changeCount++
                                connected = false
                                recorder?.stop()
                                recorder?.release()

                                Thread.sleep(500)
                                try {
                                    recorder = AudioRecord(
                                        AudioSource.DEFAULT,
                                        sampleRate,
                                        AudioFormat.CHANNEL_IN_MONO,
                                        AudioFormat.ENCODING_PCM_16BIT,
                                        recorder_buffer_size_bytes
                                    )
                                    recorder?.addOnRoutingChangedListener(this, null)
                                } catch (ex: SecurityException) {
                                    Log.d(TAG, "No audio permission")
                                    return
                                }
                                Thread.sleep(500)
                                isPeripheral = checkIfPeripheralIsConnected()
                            }
                        }
                    } else {
                        Log.d(TAG, "failed to initialize audio")
                        Thread.sleep(5000)
                    }
                }
            } catch (e: InterruptedException) {
            } finally {
                recorder?.stop()
                recorder?.release()
                player?.stop()
                player?.release()
            }
        }

        private fun playOutputSounds(read_size: Int) {
            if (clickVolume > 0.001) {
                val start_t = SystemClock.elapsedRealtime()
                var how_much_to_write = read_size * 2
                // hack to reduce amount of buffering
                // read was more than 1/10th of a second, skip some writing
                if (read_size > 4410) {
                    how_much_to_write -= 16
                }
                how_much_to_write -= 2
                var wrote_size = 0
                if (how_much_to_write > 0) {
                    wrote_size = player!!.write(
                        playbackBuffer,
                        0,
                        how_much_to_write
                    ) // , AudioTrack.WRITE_NON_BLOCKING
                }
                val end_t = SystemClock.elapsedRealtime()
                //Log.d(TAG, "Time to write: "+(end_t-start_t));
            }
        }

        private fun processInputDataAndGenerateClicks(read_size: Int, playback_data: ShortArray) {
            var i = 0
            var click_v =
                (Math.exp((clickVolume - 1.0) * Math.log(10000.0)) * 32767).toInt().toShort()
            while (i < read_size) {
                if (deadCountdown > 0) {
                    deadCountdown--
                }
                if (i * 2 + 1 < playback_data.size) {
                    if (clickCountdown > 0) {
                        clickCountdown--
                        playback_data[i * 2] = if (i % 2 == 0) click_v else 0
                        playback_data[i * 2 + 1] = playback_data[i * 2]
                    } else {
                        //playback_data[i]=0;
                        // test beep
                        //short beep=(short)((total_sample_count/40)%2 == 1 ? 1000:-1000);
                        val beep: Short = 0
                        playback_data[i * 2] = beep
                        playback_data[i * 2 + 1] = beep
                    }
                }
                val raw_v = inputBuffer[currentOffset] * (1.0f / 32768.0f)
                val v = filter.getValue(raw_v)
                if ( /* v>threshold || */v < -threshold) {
                    if (deadCountdown <= 0) {
                        totalCount++
                        sampleCount++
                        logIntervalClickCount++
                        deadCountdown = deadTime
                        if (clickCountdown <= 0) clickCountdown = clickDuration
                        appendClick(totalSampleCount)
                        trimQueue(totalSampleCount)
                    }
                }
                if (logCountdown <= 0) {
                    logCountdown = logInterval
                    countsLog.add(logIntervalClickCount)
                    logIntervalClickCount = 0
                }
                logCountdown--
                ++i
                ++currentOffset
                if (currentOffset >= inputBuffer.size) currentOffset = 0
                ++totalSampleCount
            }
        }

        override fun onRoutingChanged(audioRouting: AudioRouting) {
            Log.d(TAG, "Routing changed!")
            isPeripheral = checkIfPeripheralIsConnected()
        }

        private fun checkIfPeripheralIsConnected(): Boolean {
            if (recorder == null) return false
            var is_peripheral = false
            var checked_peripheral = false
            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    val microphones = recorder?.activeMicrophones
                    if (microphones != null) {
                        for (m in microphones) {
                            val l = m.location
                            if (l == MicrophoneInfo.LOCATION_UNKNOWN || l == MicrophoneInfo.LOCATION_PERIPHERAL) {
                                is_peripheral = true
                            }
                        }
                    }
                    checked_peripheral = true
                }
            } catch (exception: Exception) {
                Log.d(TAG, "Failed to query connected microphones")
            }
            // if we were unable to determine if microphone is peripheral, fallback to wired headset check
            if (!checked_peripheral) {
                is_peripheral =
                    (applicationContext.getSystemService(AUDIO_SERVICE) as AudioManager).isWiredHeadsetOn
            }
            return is_peripheral
        }

        private fun parametersFromConfig() {
            val prefs = PreferenceManager.getDefaultSharedPreferences(applicationContext)
            try {
                threshold = prefs.getString("threshold", "")!!.toDouble()
            } catch (e: NumberFormatException) {
            }
            try {
                deadTime =
                    (0.001 * sampleRate * prefs.getString("dead_time", "")!!.toDouble()).toInt()
            } catch (e: NumberFormatException) {
            }
            try {
                clickVolume = prefs.getInt("int_click_volume2", 100) / 100.0f
            } catch (e: NumberFormatException) {
            }
        }

        override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, s: String) {
            parametersFromConfig()
        }
    }

    @JvmField
    var listener: Listener? = null
    var listenerThread: Thread? = null

    fun start() {
        if (!started) {
            if (listenerThread == null) {
                listener = Listener()
                listenerThread = Thread(listener)
                listenerThread!!.start()
            }
            started = true
        }
    }

    fun reset() {
        totalCount = 0
        changeCount++
        lastNClicks.clear()
    }

    fun stop() {
        if (listener != null) {
            listener!!.doStop = true
            listenerThread!!.interrupt()
            try {
                listenerThread!!.join()
            } catch (e: InterruptedException) {
                // TODO Auto-generated catch block
                e.printStackTrace()
            }
            listenerThread = null
        }
        started = false
    }

    override fun onCreate() {
        super.onCreate()
        start()
    }

    override fun onTerminate() {
        super.onTerminate()
        stop()
    }

    companion object {
        private const val TAG = "MicroGeiger"
        const val running_avg_const = 0.25f
    }
}